File: snail-race-bet/python/src/domain/bet_application.py
from typing import List

from src.domain.bet import Bet
from src.domain.podium_pronostic import PodiumPronostic
from src.domain.race_result_provider import SnailRace
from src.domain.winner import Winner


class BetApplication:
    def __init__(self, bet_repository, race_result_provider):
        self.bet_repository = bet_repository
        self.race_result_provider = race_result_provider

    def place_bet(self, gambler: str, timestamp: int, first: int, second: int, third: int) -> None:
        self.bet_repository.register(Bet(gambler, PodiumPronostic(first, second, third), timestamp))

    def get_winners_for_last_race(self) -> List[Winner]:
        bets = self.bet_repository.find_by_date_range(0, float('inf'))
        races = self.race_result_provider.races()
        winning_bets = self.find_exact_match_bets(bets, races.get_last_race())

        return [Winner(bet.gambler) for bet in winning_bets]

    def find_exact_match_bets(self, bets: List[Bet], race: SnailRace) -> List['Bet']:
        return [
            bet for bet in bets
            if bet.is_in_time_for(race) and bet.bet_is_on(race.podium)
        ]

File: snail-race-bet/java/src/main/java/snail/race/kata/domain/BetApplication.java
package snail.race.kata.domain;

import java.util.List;
import java.util.stream.Stream;

public class BetApplication {
    private final BetRepository betRepository;
    private final RaceResultProvider raceResultProvider;

    BetApplication(
            BetRepository betRepository,
            RaceResultProvider raceResultProvider) {
        this.betRepository = betRepository;
        this.raceResultProvider = raceResultProvider;
    }

    void placeBet(
             String gambler,
             long timestamp,
             int first,
             int second,
             int third) {
        this.betRepository.register(new Bet(gambler, new PodiumPronostic( first, second, third), timestamp));

    }

    List<Winner> getWinnersForLastRace() {
        // Sort races by date and get the last one
        var races = raceResultProvider.races();
        var sortedRaces = races.races()
                .stream()
                .sorted((a, b) -> Long.compare(b.timestamp(), a.timestamp()))
                .toList();

        if (sortedRaces.isEmpty()) {
            return List.of();
        }

        var previousRaceTimestamp = 0;
        if( sortedRaces.size() > 1) {
            previousRaceTimestamp = Math.toIntExact(sortedRaces.get(1).timestamp());
        }

        var lastRace = sortedRaces.get(0);
        int lastRaceTimestamp = Math.toIntExact(lastRace.timestamp());
        var bets = betRepository.findByDateRange(previousRaceTimestamp, lastRaceTimestamp);

        var winningBets = findExactMatchBets(bets, lastRace);

        return winningBets.map(bet -> new Winner(bet.gambler())).toList()   ;

    }

    private Stream<Bet> findExactMatchBets(List<Bet> bets, RaceResultProvider.SnailRace race) {
        return bets.stream()
                .filter(bet -> bet.isInTimeFor(race))
                .filter(bet -> bet.betIsOn(race.podium()));
    }

}
